script/analysis/torus_id.py

```diff
@@ -0,0 +1,244 @@
+    Z1 = 1 + (1-a*a)**(1./3.) * ((1 + a)**(1./3.) + (1 - a)**(1./3.))
+    Z2 = np.sqrt(3*a*a + Z1*Z1)
+    return 3 + Z2 - np.sign(a)*np.sqrt((3-Z1)*(3+Z1+2*Z2))
+
+def lfish(a, r):
+    """Get the specific angular momentum
+    from Fishbone and Moncrief, 1976."""
+    from numpy import sqrt
+    try:
+        return (
+      ((pow(a, 2) - 2. * a * sqrt(r) + pow(r, 2)) *
+          ((-2. * a * r * (pow(a, 2) - 2. * a * sqrt(r) + pow(r, 2))) /
+                  sqrt(2. * a * sqrt(r) + (-3. + r) * r) +
+              ((a + (-2. + r) * sqrt(r)) * (pow(r, 3) + pow(a, 2) * (2. + r))) /
+                  sqrt(1 + (2. * a) / pow(r, 1.5) - 3. / r))) /
+      (pow(r, 3) * sqrt(2. * a * sqrt(r) + (-3. + r) * r) *
+          (pow(a, 2) + (-2. + r) * r)))
+    except: 
+        return 0
+
+def get_lnh(a, rin, rmax, r, th):
+    """Get the log of the specific enthalpy 
+    from Fishbone, Moncrief, 1976."""
+    # Assumes G = c = M = 1
+    from numpy import log, sqrt
+    
+    l = lfish(a, rmax)
+    sth = np.sin(th)
+    cth = np.cos(th)
+    
+    DD = r * r - 2. * r + a * a
+    AA = ((r * r + a * a) * (r * r + a * a) - DD * a * a * sth * sth)
+    SS = r * r + a * a * cth * cth
+    
+    thin = np.pi/2.
+    sthin = np.sin(thin)
+    cthin = np.cos(thin)
+    DDin  = rin * rin - 2. * rin + a * a;
+    AAin  = ((rin * rin + a * a) * (rin * rin + a * a) 
+             - DDin * a * a * sthin * sthin)
+    SSin  = rin * rin + a * a * cthin * cthin
+    
+    try:
+        return (0.5 * log((1. + sqrt(1. + 4. * (l * l * SS * SS) * DD /
+                                        (AA * sth * AA * sth))) /
+                    (SS * DD / AA)) -
+          0.5 * sqrt(1. + 4. * (l * l * SS * SS) * DD / (AA * AA * sth * sth)) -
+          2. * a * r * l / AA -
+          (0.5 * log((1. + sqrt(1. + 4. * (l * l * SSin * SSin) * DDin /
+                                         (AAin * AAin * sthin * sthin))) /
+                     (SSin * DDin / AAin)) -
+              0.5 * sqrt(1. + 4. * (l * l * SSin * SSin) * DDin /
+                                  (AAin * AAin * sthin * sthin)) -
+              2. * a * rin * l / AAin))
+    except:
+        return 0
+
+def rho_from_hm1_ideal(hm1):
+    "Dummy EOS call"
+    kappa = 2e-7
+    gam = 4./3.
+    rho = pow(hm1 * (gam - 1.) / (kappa * gam), 1. / (gam - 1.))
+    return rho
+
+def compute_rho_from_coords(a, rin, rmax, eos, r, th):
+    "Get the density from Fishbone, Moncrief, 1976."
+    lnh = get_lnh(a, rin, rmax, r, th)
+    if type(lnh) == np.ndarray:
+        lnh[np.isnan(lnh)] = 0.
+        lnh[r <= rin] = 0.
+        lnh[lnh <= 0] = 0.
+    elif np.isnan(lnh) or r <= rin or lnh <= 0:
+        lnh = 0.
+    hm1 = np.exp(lnh) - 1
+    return eos(hm1)
+
+def get_tot_mass(M, a, rin, rmax, eos, get_max = False):
+    """Integrate the mass in a fishbone-moncrief disk.
+    Returns mass in solar masses."""
+    NX1 = 512
+    NX2 = 512
+    X1 = np.linspace(np.log(rin),np.log(1e3), NX1)
```

Comment:
Worth checking for rmax relative to 1e3? e.g. `np.log(max(1e3, 2*rmax))`?

Reply:
Good call. I'll add this.

Reply:
Sorry, not `rmax` since that's just the pressure max radius right. I forget if there's an easy way to get rout. You could check that the fishbone solution is 0 (or NaN or whatever) at r = 1e3